### ISSUE TITLE
Where / how to store persistent files from the docker image

### ISSUE THREAD

**Issue by 4shaw** [arunoda/meteor-up-legacy#454]
Hi... Im trying to save persistent files that can be served via nginx.
I cant save to the public directory as this is always overwritten on deploy.
Im trying to write to the file system with: 
fs.writeFile('static/<path>', ...)
this lands up on the host in: /opt/appname/current/bundle/programs/server/static
This gets overwritten on every deploy
then i tried:
fs.writeFile('/static/path', ...) Note the preceeding '/'
Im not sure where these files are... probably in the docker container?

Is there no volume on the container to store images etc that will persist through the deployment process
How would you go about doing this?
Thanks

Possible solution could be to add a volume for these files on the meteorhacks/meteord:base image?
Currently:
"Volumes": {
        "/bundle": "/opt/appname/current"
},
To:
"Volumes": {
        "/bundle": "/opt/appname/current",
        "/static": "/opt/appname/static"
},

**Comment by hems** [comment]
@4shaw this is something i would like to understand as well!
- 1

in the meanwhile, if you just saving images an interesting workaround might be to use http://cloudinary.com which would host them in a CDN and give you transformations via url.

i found this meteor package that might be useful: https://github.com/Lepozepo/cloudinary

**Comment by mmoyles87** [comment]
I've just been using S3 personally. It's very inexpensive, and cloud storage seems to jive well with the docker philosophy.

On Sat, May 30, 2015 at 4:40 PM, henrique matias notifications@github.com
wrote:

> @4shaw https://github.com/4shaw this is something i would like to
> understand as well!
> - 1
> 
> in the meanwhile, if you just saving images and interesting workaround
> might be to use http://cloudinary.com which would host them in a CDN and
> give you transformations via url.
> 
> i found this meteor package that might be useful:
> https://github.com/Lepozepo/cloudinary
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/arunoda/meteor-up/issues/454#issuecomment-107082670.

## 

Matt Moyles
http://mattmoyles.com

**Comment by hems** [comment]
@dubvfan87 indeed, but sometimes you sont want to trust your files to other hosts ( :

**Comment by 4shaw** [comment]
Thanks for the suggestions.
I agree with hems that there needs to be a solution local storage too.

I have tested by downloading the mupx repo and adding the volume for storage and it works. I called it "static"

The new directory structure at /opt/-appname-/ looks like:

config/
current/
last/
static/
tmp/

The static/ directory has been added and is not affected during deployment process... from within meteor you can access the static directory as though it were a root directory:
/static/-subfolder-/image.jpg

Then with nginx you can create an alias to point to to your static directory at:
/opt/-appname-/static

It would be cool if the meteor up team could consider adding this to the mupx branch so that we could start using it.

**Comment by jowilhnson** [comment]
Are there any other solutions to allow us to write to the filesystem? We have meteor apps running on a server in our intranet, one of which after updating required mupx to get redeployed. It builds PDFs and writes them to a shared folder in the filesystem. Or it used to anyhow, now it throws an "Error: ENOENT, open '/..../...pdf'" exception. 

Editing the mupx repo, installing/configuring nginx etc is all a bit much for us just to be able to write files to the filesystem, so hoping there's a simpler workaround.

**Comment by MasterJames** [comment]
I'm thinking you could mount a volume with a change to the docker run command.

Here is an interesting thread as well.
https://github.com/arunoda/meteor-up/issues/458

It's this start.sh file that is where docker container is run with other volume mount entries.
https://github.com/arunoda/meteor-up/blob/mupx/templates/linux/start.sh

You could also I imagine use docker exec to execute the nessecary commands after mupx is done from ssh possibly.

The new port underway at kadirahq may have an option for that? Or a request there might be warranted.

**Comment by zb0t** [comment]
@4shaw I know this is a bit late but I'd like to try your method. Where can I add the additional volumes to add? I'd like to replicate your "static" directory structure

**Comment by 4shaw** [comment]
Hi... sorry for the delayed reply

I can't remember everything I did to get it to work but I have the files on
Bitbucket... you use it exactly the same as mupx but with _mup4_ instead

eg:
mup4 deploy
etc....

You can clone the repo with the below url:
https://b4shaw@bitbucket.org/b4shaw/mup4.git

I just install mup4 from the directory like this:

npm install -g <foldernam of the mup4 repo on your machine>

its outdated now im sure but it still work well for me

On Sun, Jul 31, 2016 at 9:18 PM, Zachary Botterman <notifications@github.com

**Comment by popaprozac** [comment]
@4shaw No problem! I deleted my comment because I found a sustainable solution for me.

Found mupx directory in `/usr/local/lib/node_modules/mupx` and modified the following:

`templates/linux/start.sh` added:
`--volume=$APP_PATH: / \`

`scripts/linux/setup-env.sh` added:
`sudo mkdir -p /opt/<%= appName %>/static`

I had to rerun `mupx setup` and `deploy`. 

The static directory is now accessible and persistent!

**Comment by greenwolfe** [comment]
zb0t, can you explain a little more about what you did? How does your volume command end up pointing to your static directory?  Where/how is APP_PATH set, in your settings.json file?  I assume you could make your static directory manually, but found it easier to automate it?

THanks,

Matt

**Comment by popaprozac** [comment]
@greenwolfe I have a typo!
It should be:
`--volume=$APP_PATH:/home \`

It is accessible at `/home/static`.

I had issues not adding a base director, i.e. `home`. So that is why you see that prefixing `static`.

Notice in `templates/linux/start.sh` the top two lines:

`APPNAME=<%= appName %>
APP_PATH=/opt/$APPNAME`

You could make the directory manually but in case I change my deployment location, this will be setup correctly every time.

Hope this helps.

**Comment by greenwolfe** [comment]
Thanks for the help! I got it working with the help also of [issue235](https://github.com/tomitrescak/meteor-uploads/issues/235) from the meteor-uploads package.

**Comment by romaroma** [comment]
mup now supports 'volumes' option in mup.js
It seems to be a solution to the problem.
Just make a directory on your server, say /opt/APPNAME/uploads and map it to some directory in your docker - /opt/uploads or so.

**Comment by greenwolfe** [comment]
Still using meteor@1.1 and plan to stay that way for this year until things settle down some more.  There are only so many times I can afford to completely rewrite my code.  When I tried the lastest mup, it did not work with my version of meteor, so mupx with the hack for volumes is my best option until next summer.
